uCNC/src/core/motion_control.c

```diff
@@ -1150,25 +1150,26 @@ uint8_t mc_build_hmap(float *target, float *offset, float retract_h, motion_data
 	float offset_y = offset[1] / (H_MAPING_GRID_FACTOR - 1);
 	float position[AXIS_COUNT];
 	float feed = block_data->feed;
-
+	
 	mc_get_position(position);
+	
+	float minretract_h = position[AXIS_TOOL] + retract_h;
 
 	for (uint8_t j = 0; j < H_MAPING_GRID_FACTOR; j++)
 	{
 		target[AXIS_X] = start_x;
 		for (uint8_t i = 0; i < H_MAPING_GRID_FACTOR; i++)
 		{
 			block_data->feed = FLT_MAX;
-			// retract if needed
-			if (position[AXIS_TOOL] < (target[AXIS_TOOL] + retract_h))
+			// retract (higher if needed)
+			retract_h = MAX(minretract_h, position[AXIS_TOOL] + retract_h);
+			position[AXIS_TOOL] = retract_h;
```

Comment:
This adds the `position[AXIS_TOOL]` to `retract_h` on every loop, so the retraction ends up far too high.

Maybe:

```cpp
minretract_h = MAX(minretract_h, position[AXIS_TOOL] + retract_h);
position[AXIS_TOOL] = minretract_h;
```

Reply:
I added a second retract variable to keep track of the highest retraction height.
The idea is to prevent collision while traveling from the last to the initial point in the hmap.

I will take a second look at the code. I'm not seeing the G39.1 ang G39.2 implementations